lib/Parse/ParseExpr.cpp

```diff
@@ -1418,35 +1442,23 @@ ParserResult<Expr> Parser::parseExprPostfix(Diag<> ID, bool isExprBasic) {
         continue;
       }
 
-      // Handle "x.dynamicType" - A metatype expr.
-      // Deprecated in SE-0096: `x.dynamicType` becomes `type(of: x)`
-      //
-      // FIXME(SE-0096): This will go away along with the keyword soon.
-      if (Tok.is(tok::kw_dynamicType)) {
-        // Fix-it
+      // Handle "x.self" expr.
+      if (Tok.is(tok::kw_self)) {
+        Result = makeParserResult(
+          new (Context) DotSelfExpr(Result.get(), TokLoc, consumeToken()));
+        continue;
+      }
+
+      // Handle the deprecated 'x.dynamicType' and migrate it to `type(of: x)`
+      if (Tok.getText() == "dynamicType") {
         auto range = Result.get()->getSourceRange();
         auto dynamicTypeExprRange = SourceRange(TokLoc, consumeToken());
         diagnose(TokLoc, diag::expr_dynamictype_deprecated)
           .highlight(dynamicTypeExprRange)
           .fixItReplace(dynamicTypeExprRange, ")")
           .fixItInsert(range.Start, "type(of: ");
-
-        // HACK: Arbitrary.
-        auto loc = range.Start;
-        auto dt = new (Context) DynamicTypeExpr(loc, loc, Result.get(), loc, Type());
-        dt->setImplicit();
-        Result = makeParserResult(dt);
```

Comment:
IMO, we should parse and build AST as possible as we can.
I tried this PR locally:
```swift
func foo<T>(_: T.Type) {}
let x = 1
foo(x.dynamicType)
```
result:
```
test3888.swift:3:6: error: '.dynamicType' is deprecated. Use 'type(of: ...)' instead
foo(x.dynamicType)
     ^~~~~~~~~~~~
    type(of:  )
test3888.swift:3:1: error: cannot invoke 'foo' with an argument list of type '(Int)'
foo(x.dynamicType)
^
test3888.swift:3:1: note: expected an argument list of type '(T.Type)'
foo(x.dynamicType)
^
```
In this case, I think, we should not emit the second error.
Since `x.dynamicType` doesn't form `DynamicTypeExpr `, but just `UnresolvedDeclRefExpr`, TypeChecker cannot know the intended type of `x.dynamicType`.